hashicorp/waypoint
# plugin/docker+k8s: using resource manager to fill in declared resources

### What's changing

This change uses the DeclaredResources wiring introduced in https://github.com/hashicorp/waypoint/pull/1737, and depends on this waypoint-plugin-sdk change: https://github.com/hashicorp/waypoint-plugin-sdk/pull/45

In this change, the Kubernetes and Docker plugins are exposing their resources that are directly under management to the waypoint server. 

### How to verify

Run a deployment using docker or k8s (i.e. run `waypoint up` in this [example project](https://github.com/hashicorp/waypoint-examples/tree/main/kubernetes/nodejs)), then look for a new `declaredResource` section in the response for the GetDeployment api. You can do this with grpcurl:

```
SEQ_NUM=1; APP=example-nodejs; PROJECT=$APP; \
grpcurl -d '{"ref": {"sequence": { "application": {"application": "'$APP'", "project": "'$PROJECT'"}, "number": '$SEQ_NUM' }}, "load_details": "0"}' \
--insecure -H "client-api-protocol: 1,1" -H "Authorization: `waypoint token new`" \
localhost:9701 hashicorp.waypoint.Waypoint/GetDeployment
```

And you'll see a section like this in the response:

Kubernetes
```
  "declaredResources": [
    {
      "id": "01FB4QN6S4H2RZYR88VTHXYM4J",
      "name": "deployment",
      "platform": "kubernetes",
      "state": {
        "@type": "type.googleapis.com/k8s.Resource.Deployment",
        "name": "example-nodejs-k8s-01fb4qmq0rf4gnd4871v0kqdwt"
      },
      "stateJson": "{\"name\":\"example-nodejs-k8s-01fb4qmq0rf4gnd4871v0kqdwt\"}",
      "categoryDisplayHint": "INSTANCE_MANAGER"
    }
  ]
```

Docker:
```
  "declaredResources": [
    {
      "id": "01FB4TKE5ZW6KSW6ZF66D0MZ4F",
      "name": "network",
      "platform": "docker",
      "state": {
        "@type": "type.googleapis.com/docker.Resource.Network",
        "name": "waypoint"
      },
      "stateJson": "{\"name\":\"waypoint\"}",
      "categoryDisplayHint": "ROUTER"
    },
    {
      "id": "01FB4TKE5ZW6KSW6ZF66D0MZ4G",
      "name": "container",
      "platform": "docker",
      "state": {
        "@type": "type.googleapis.com/docker.Resource.Container",
        "id": "66bb53c2df7818210e7e05302107ac98b3457cb66d690d90e2fde6ce32950130",
        "name": "web-01FB4TKDC98T4MYSFXBZPYG5RE"
      },
      "stateJson": "{\"id\":\"66bb53c2df7818210e7e05302107ac98b3457cb66d690d90e2fde6ce32950130\",\"name\":\"web-01FB4TKDC98T4MYSFXBZPYG5RE\"}",
      "categoryDisplayHint": "INSTANCE"
    }
  ]
```
